src/diagnostics/lintstore.md

```diff
@@ -0,0 +1,40 @@
+# LintStore
+
+This page documents some of the machinery around lint registration and how we
+run lints in the compiler.
+
+The `LintStore` is the central piece of infrastructure, around which everything
+rotates. It's not available during the early parts of compilation (i.e., before
+TyCtxt) in most of the code, as we need to fill it in with all of the lints,
+which can only happen after plugin registration.
+
+We store essentially two separate things in the `LintStore`: a list of all known
```

Comment:
I think I would move the `## Registration` section up here -- basically say:

> During the mutable period, we can register lints by providing a "constructor function", like so:

This existing paragraph says "We store constructors for the lint passes rather than the passes
themselves" but this feels (to me) like something that makes more sense if you happen to be familiar with the old code than from first principles. i.e., there is nothing all that strange about the fact that we register a function that can produce a lint (also, it's not always a *function* right? it's basically any closure?)

So I would just describe it as is: to register the lint, you give metadata + a constructor. When it is time for us to run the lints, we invoke the closure and get back a `Box<dyn LintPass>` value. We can then invoke the visitor methods on this lint pass. Note that the lint pass methods get `&mut self` so the lint can keep internal, mutable state that it updates as we traverse the input program.